src/en/RoyalRoad.lua

```diff
@@ -15,23 +128,87 @@ local function expandURL(url)
 	return baseURL .. (url:sub(1, 1) == "/" and "" or "/") .. url
 end
 
+local function TriStateFilter_(int, str)
+	if TriStateFilter then
+		return TriStateFilter(int, str)
+	else
+		return CheckboxFilter(int, str)
+	end
+end
+
+local function MultiTriStateFilter(offset, filter_ext, stop)
+	local f={}
+	for int = 1,stop,1
+	do
+		table.insert(f, TriStateFilter_(offset+int, filter_ext[int]))
+	end
+	return f
+end
+
+local function TriQuery(data, filter_int, int)
+	return (data[int] and (data[int]==2 and "&tagsRemove" or "&tagsAdd=")..filter_int[int] or "")
+end
+
+local function MultiTriQuery(data, filter_int, start, stop)
+	local q=""
+	for int = start,stop,1
+	do
+		q =q.. TriQuery(data, filter_int, int)
+	end
+	return q
+end
+
+local function createFilterString(data)
+	return string.gsub("?"..
+		(data[QUERY] and "&title="..data[QUERY] or "")..
+		(data[KEYWORD_FILTER_KEY]~="" and "&keyword="..data[KEYWORD_FILTER_KEY] or "")..
+		(data[AUTHOR_FILTER_KEY]~="" and "&author="..data[AUTHOR_FILTER_KEY] or "")..
+		MultiTriQuery(data, GENRES_FILTER_INT, 201, 215)..
+		MultiTriQuery(data, TAGS_FILTER_INT, 301, 346)..
+		MultiTriQuery(data, CONTENT_WARNINGS_FILTER_INT, 401, 404)..
+		(data[PAGES_MIN_FILTER_KEY ]~="" and "&minPages=" ..data[PAGES_MIN_FILTER_KEY ] or "")..
+		(data[PAGES_MAX_FILTER_KEY ]~="" and "&maxPages=" ..data[PAGES_MAX_FILTER_KEY ] or "")..
+		(data[RATING_MIN_FILTER_KEY]~="" and "&minRating="..data[RATING_MIN_FILTER_KEY] or "")..
+		(data[RATING_MAX_FILTER_KEY]~="" and "&maxRating="..data[RATING_MAX_FILTER_KEY] or "")..
+		(data[601] and "&status="..STATUS_FILTER_INT[601] or "")..
+		(data[602] and "&status="..STATUS_FILTER_INT[602] or "")..
+		(data[603] and "&status="..STATUS_FILTER_INT[603] or "")..
+		(data[604] and "&status="..STATUS_FILTER_INT[604] or "")..
+		(data[605] and "&status="..STATUS_FILTER_INT[605] or "")..
+		(data[606] and "&status="..STATUS_FILTER_INT[606] or "")..
+		(data[ORDER_BY_FILTER_KEY]~=0 and "&orderBy="..ORDER_BY_FILTER_INT[data[ORDER_BY_FILTER_KEY]] or "")..
+		(data[ORDER_FILTER_KEY] and "&dir=asc" or "")..
+		(data[TYPE_FILTER_KEY]~=0 and "&type="..TYPE_FILTER_INT[data[TYPE_FILTER_KEY]] or "")
+	, "?&", "?")
```

Comment:
1. Avoid excessive use of `and`/`or`. It can be useful a few times but this much is abusive.
2. Use the `url` library's `querystring`, which vastly simplifies the code. This also solves problem 1.
3. Avoid repetition of magic constants.

```suggestion
	local function emptyNil(str)
		if str == "" then
			return nil
		end
		return str
	end

	local statuses = {}
	for i=1,#STATUS_FILTER_EXT do
		if data[STATUS_FILTER_KEY+i] then
			statuses[#statuses+1] = STATUS_FILTER_INT[STATUS_FILTER_KEY+i]
		end
	end
	
	local order
	if data[ORDER_BY_FILTER_KEY] ~= 0 then
		order = ORDER_BY_FILTER_INT[data[ORDER_BY_FILTER_KEY]]
	end
	
	local dir
	if data[ORDER_FILTER_KEY] then
		dir = "asc"
	end
	
	local type
	if data[TYPE_FILTER_KEY] ~= 0 then
		type = TYPE_FILTER_INT[data[TYPE_FILTER_KEY]]
	end
	
	local tagsAdd, tagsRemove = {}, {}
	local function MultiTryQuery(strings, start, len)
		for i=start+1,start+len do
			if data[i] then
				if data[i] == 2 then
					tagsRemove[#tagsRemove+1] = strings[i]
				else
					tagsAdd[#tagsAdd+1] = strings[i]
				end
			end
		end
	end
	
	MultiTriQuery(GENRES_FILTER_INT, GENRES_FILTER_KEY, #GENRES_FILTER_EXT)
	MultiTriQuery(TAGS_FILTER_INT, TAGS_FILTER_KEY, #TAGS_FILTER_EXT)
	MultiTriQuery(CONTENT_WARNINGS_FILTER_INT, CONTENT_WARNINGS_FILTER_KEY, #CONTENT_WARNINGS_FILTER_EXT)
	
	return "?"..qs({
		title = data[QUERY],
		keyword = emptyNil(data[KEYWORD_FILTER_KEY]),
		author = emptyNil(data[AUTHOR_FILTER_KEY]),
		minPages = emptyNil(data[PAGES_MIN_FILTER_KEY]),
		maxPages = emptyNil(data[PAGES_MAX_FILTER_KEY]),
		minRating = emptyNil(data[RATING_MIN_FILTER_KEY]),
		maxRating = emptyNil(data[RATING_MAX_FILTER_KEY]),
		status = statuses,
		orderBy = order,
		dir = dir,
		tagsAdd = tagsAdd,
		tagsRemove = tagsRemove
	})
```

Reply:
Note: as you may see in my suggested change, **readability should be preferred over conciseness**. This is especially the case when readability can be drastically improved with no loss of functionality.